File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque

def bfs(start_x, start_y, end_x, end_y, graph):
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    n, m = len(graph), len(graph[0])

    visited = {(start_x, start_y)}  # 방문 집합 생성
    queue = deque([(start_x, start_y, 1)])  # 큐에 (x, y, 거리) 넣기

    while queue:
        x, y, dist = queue.popleft()

        if x == end_x and y == end_y:  # 탐색 종료
            return dist

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < n and 0 <= ny < m: # 배열 범위 내에 있는지
                if (nx, ny) not in visited:
                    if graph[nx][ny] == 1: # 갈 수 있다면
                        visited.add((nx, ny))
                        queue.append((nx, ny, dist + 1))  # (다음좌표x, 다음좌표y, 현재거리 + 1)

    return -1  # 큐가 비었는데 도달 못했을 경우


N, M = map(int, sys.stdin.readline().split())
graph = [[0] * N for _ in range(N)]
for i in range(N):
    graph[i] = list(map(int, sys.stdin.readline().rstrip()))

print(bfs(0,0,N-1,M-1,graph))

File: 백준/Silver/2178. 미로 탐색/미로 탐색.cc
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

/*

*/

int dy[] = {0, 0, 1, -1};
int dx[] = {1, -1, 0, 0};

int N, M;
int graph[100][100];
int dist[100][100];
bool visited[100][100];

int bfs(int start_x, int start_y){
    queue<pair<int,int>> q;
    q.push({start_x,start_y});
    visited[start_x][start_y]=true;
    
    while (!q.empty()){
        int x = q.front().first;
        int y = q.front().second;
        int now_dist = dist[x][y];
        
        if (x==N-1 && y==M-1){
            return dist[N-1][M-1];
        }

        q.pop();
        for (int i=0; i<4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if ( nx<0 || nx>=N || ny<0 || ny>=M ) continue;
            if (graph[nx][ny]==0) continue;
            if (visited[nx][ny]) continue;
            visited[nx][ny] = true;
            dist[nx][ny] = now_dist + 1;
            q.push({nx, ny});
        }
    }
    return -1;  // 도달 못했을 경우
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    cin >> N >> M;
    for (int i=0; i<N; i++){
        for (int j=0; j<M; j++){
            char getChar;
            cin >> getChar;
            graph[i][j] = getChar - '0';
        }
    }

    // 이 문제에서는 시작 위치도 답에 포함하므로 +1
    cout << bfs(0,0) + 1;

    return 0;
}